[processing] restore GDAL algorithms

alexbruy opened the pull request:
## Description
Restore and homogenize most of GDAL algorithms. Where possible extent parameter reprojected into input layer CRS, to avoid errors and freezes.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

alexbruy committed:
[processing] restore pct2rgb algorithm

alexbruy committed:
fix permissions

alexbruy committed:
[processing] restore polygonize algorithm

Also expose band parameter

alexbruy committed:
[processing] restore TPI algorithm

alexbruy committed:
[processing] restore slope algorithm

alexbruy committed:
[processing] restore roughness algorithm

alexbruy committed:
[processing] use native API to get GDAL driver name from filename

alexbruy committed:
remove duplicate tests

alexbruy committed:
[processing] restore contour algorithm

Also expose new parameters and use custom widget for additional
creation options

alexbruy committed:
[processing] restore sieve algorithm

alexbruy committed:
[processing] restore fillnodata algorithm

alexbruy committed:
[processing] minor fixes in rgb2pct algorithm

alexbruy committed:
[processing] restore gdaltindex algorithm and expose additional parameters

alexbruy committed:
[processing] restore moving average interpolation

alexbruy committed:
[processing] restore grid data metrics algorithm

alexbruy committed:
[processing] restore inverse distance to a power algorithm

alexbruy committed:
[processing] restore nearest neighbor interpolation

alexbruy committed:
[processing][FEATURE] add IDW with neareast neighbor search

alexbruy committed:
[processing][FEATURE] add linear interpolation

alexbruy committed:
[processing] restore clip raster by extent algorithm

alexbruy committed:
[processing] restore clip raster by mask algorithm

alexbruy committed:
[processing] restore proximity algorithm

alexbruy committed:
[processing] homogenize gdal algorithms

-consistent parameters naming
-use additional creation options where possible
-remove duplicated functionality

alexbruy committed:
[processing] restore gdal2tiles algorithm

alexbruy committed:
fix optional CRS parameter handling

alexbruy committed:
[processing] restore retile algorithm

alexbruy committed:
[processing] restore gdal2xyz algorithm

alexbruy committed:
[processing] restore gdaladdo algorithm

alexbruy committed:
[processing] rename some algorithms to avoid naming conflicts and
maintain consistency

alexbruy committed:
[processing] restore ogrinfo algorithm

alexbruy committed:
[processing] restore offset curve algorithm

alexbruy committed:
fix typos and add tests for gdal algorithms

alexbruy committed:
[processing] fixes to Offset curve algorithm

alexbruy committed:
[processing] restore buffer algorithm

alexbruy committed:
[processing] homogenize and improve Points along lines

alexbruy committed:
[processing] restore one side buffer algorithm

alexbruy committed:
[processing] restore clip vector by extent algorithm

alexbruy committed:
[processing] restore clip vector by mask algorithm

alexbruy committed:
remove unused includes

alexbruy committed:
[processing] restore ogr2ogr algorithm

alexbruy committed:
[processing] restore dissolve algorithm

alexbruy committed:
[processing] restore Execure SQL algorithm

alexbruy committed:
[processing] restore gdal_merge algorithm

alexbruy committed:
[processing] tests for OGR-based algorithms

nyalldawson commented on python/plugins/processing/algs/gdal/ExecuteSql.py:
```diff
@@ -0,0 +1,117 @@
+        self.addParameter(QgsProcessingParameterString(self.SQL,
+                                                       self.tr('SQL expression'),
+                                                       defaultValue=''))
+        self.addParameter(QgsProcessingParameterEnum(self.SQL,
```
Looks like a copy/paste mistake here?

nyalldawson commented on python/plugins/processing/algs/gdal/GridAverage.py:
```diff
@@ -51,70 +53,122 @@ class GridAverage(GdalAlgorithm):
-                                             self.tr('Output raster type'), self.TYPE, 5))
-
-        self.addOutput(OutputRaster(self.OUTPUT, self.tr('Interpolated moving average')))
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Feature source here?

nyalldawson commented on python/plugins/processing/algs/gdal/GridDataMetrics.py:
```diff
@@ -53,39 +55,87 @@ class GridDataMetrics(GdalAlgorithm):
+                        (self.tr('Average distance'), 'average_distance'),
+                        (self.tr('Average distance between points'), 'average_distance_pts'))
+
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Feature source?

nyalldawson commented on python/plugins/processing/algs/gdal/GridInverseDistance.py:
```diff
@@ -0,0 +1,197 @@
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
feature source?

nyalldawson commented on python/plugins/processing/algs/gdal/GridInverseDistanceNearestNeighbor.py:
```diff
@@ -0,0 +1,182 @@
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Feature source?

nyalldawson commented on python/plugins/processing/algs/gdal/GridLinear.py:
```diff
@@ -0,0 +1,153 @@
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Feature source

nyalldawson commented on python/plugins/processing/algs/gdal/GridLinear.py:
```diff
@@ -0,0 +1,153 @@
+        z_field_param.setFlags(z_field_param.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
+        self.addParameter(z_field_param)
+
+        self.addParameter(QgsProcessingParameterNumber(self.RADIUS,
```
What's a -1 search distance represent here?

alexbruy replied:
From GDAL docs
>If set to -1, the search distance is infinite. If set to 0, nodata value will be always used.

nyalldawson commented on python/plugins/processing/algs/gdal/GridNearestNeighbor.py:
```diff
@@ -0,0 +1,168 @@
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
feature source?

nyalldawson commented:
Wow, fantastic work!

alexbruy committed:
[processing] fix import

alexbruy committed:
[processing] fix copy-paste error in Execute SQL

alexbruy committed:
[processing] use feature source in the interpolation algorithms

alexbruy committed:
[processing] reproject extent into input CRS where possible

alexbruy committed:
[processing] tests for clipping (raster and vector) algorithms

alexbruy commented on python/plugins/processing/algs/gdal/ClipRasterByMask.py:
```diff
@@ -0,0 +1,164 @@
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterRasterLayer(self.INPUT,
+                                                            self.tr('Input layer')))
+        self.addParameter(QgsProcessingParameterVectorLayer(self.MASK,
```
Hmm, not sure that for this algorithm it makes sense. I will investigate

alexbruy committed:
fix indentation

alexbruy committed:
[processing] fix GDAL raster tests

alexbruy committed:
[processing] fix SQL-based geoprocessing tools and tests

alexbruy committed:
[processing] fix clip by extent tests

alexbruy committed:
[processing] fix test for clip vector by mask algorithms

alexbruy committed:
[processing] fix execute sql test

alexbruy committed:
[processing][API]introduce QgsProcessingOutputFile output and fix gdal2xyz test

alexbruy commented:
Tests passed locally, but failed on Travis (probably because different GDAL versions/builds are used).

nirvn commented:
@alexbruy , anything holding this nice PR off?

alexbruy commented:
@nirvn, only failing tests. They all pass for me locally, but fail on Travis. I just rebased to have latest changes (using docker for testing) and still no luck.

alexbruy committed:
[processing] disable clip by polygon tests as they pass locally but
fail on Travis with TopologyException

ERROR 1: TopologyException: Input geom 1 is invalid: Self-intersection
at or near point 2 -1 at 2 -1

alexbruy committed:
[processing] exclude location field from check

alexbruy committed:
[processing] disable test for script-based algorithms as GDAL scripts
are not available on Travis

alexbruy committed:
[processing] add second hash for linear grid interpolation test

alexbruy commented:
I disabled tests which failed because of missed GDAL Python scrips (e.g. `gdal2xyz.py`, `gdal_polygonize.py` etc) and now all test passed. The only "failing" test is spellcheck, it timed out probably because of massive update of expected files. IMHO it is safe to ignore this failure and merge.

@nyalldawson, @m-kuhn are you ok with merging this despite failing spellcheck test?

nyalldawson commented:
No problem merging with spell check timeout... It'll fix itself after merge.

Great work on this!!

nirvn commented:
Woupidou!
